File: sorting/quicksort/Quicksort.java
package quicksort;

import java.util.Arrays;
import java.util.Random;

public class Quicksort {

    public static void main(String[] args) {
        int[] arr = {10, 7, 8, 9, 1, 5};
        quicksort(arr, 0, arr.length - 1);
        System.out.println(Arrays.toString(arr));
    }

    public static void quicksort(int[] arr, int low, int high) {
        if (low < high) {
            int pi = partition(arr, low, high);

            quicksort(arr, low, pi - 1);
            quicksort(arr, pi + 1, high);
        }
    }

    static int partition(int[] arr, int low, int high) {
        Random random = new Random();
        int pivotIndex = low + random.nextInt(high - low + 1);
        int pivot = arr[pivotIndex];

        // Move pivot to the end
        swap(arr, pivotIndex, high);

        int i = (low - 1); // index of smaller element
        for (int j = low; j < high; j++) {
            // If current element is smaller than or
            // equal to pivot
            if (arr[j] <= pivot) {
                i++;

                // swap arr[i] and arr[j]
                swap(arr, i, j);
            }
        }

        // swap arr[i+1] and arr[high] (or pivot)
        swap(arr, i + 1, high);

        return i + 1;
    }

    // A utility function to swap two elements
    static void swap(int[] arr, int i, int j) {
        int temp = arr[i];
        arr[i] = arr[j];
        arr[j] = temp;
    }
}

File: sorting/quicksort/quicksort.py
from random import choice


def quicksort(arr: list[int]) -> list[int]:
    if len(arr) < 2:
        return arr
    pivot = choice(arr)
    lowers = [i for i in arr if i <= pivot]
    equals = [i for i in arr if i == pivot]
    highers = [i for i in arr if i > pivot]

    return quicksort(lowers) + equals + quicksort(highers)


if __name__ == "__main__":
    base_arr = [10, 45, 13, 17, 29, 90]

    print(f"base array => {base_arr}\nordered array => {quicksort(base_arr)}")
